examples/seismic/source.py

```diff
@@ -1,60 +1,105 @@
 from devito.logger import error
 
+from cached_property import cached_property
+
 import numpy as np
 try:
     import matplotlib.pyplot as plt
 except:
     plt = None
 
-__all__ = ['PointSource', 'Receiver', 'Shot', 'RickerSource', 'GaborSource']
+__all__ = ['PointSource', 'Receiver', 'Shot', 'WaveletSource',
+           'RickerSource', 'GaborSource', 'TimeAxis']
+
+
+class TimeAxis:
+    def __init__(self, start, step, num):
+        try:
+            self.start = start
+            self.stop = start + step*num
+            self.step = step
+            self.num = num
+        except:
+            error("error in time range parameters")
+
+    def __str__(self):
+        return "TimeAxis(%g, %g, %g, %g)" % (self.start, self.stop, self.step, self.num)
+
+    @cached_property
+    def time_values(self):
+        return np.linspace(self.start, self.stop, self.num, endpoint=False)
 
 
 class PointSource(SparseTimeFunction):
     """Symbolic data object for a set of sparse point sources
 
-    :param name: Name of the symbol representing this source
+    :param name: Name of the symbol representing this source.
     :param grid: :class:`Grid` object defining the computational domain.
-    :param coordinates: Point coordinates for this source
-    :param data: (Optional) Data values to initialise point data
-    :param ntime: (Optional) Number of timesteps for which to allocate data
-    :param npoint: (Optional) Number of sparse points represented by this source
-    :param time_order: (Optional) Time discretization order (defaults to 2)
+    :param coordinates: Point coordinates for this source.
+    :param time_range: TimeAxis(start, step, num) object.
+    :param data: (Optional) Data values to initialise point data.
+    :param npoint: (Optional) Number of sparse points represented by this source.
+    :param time_order: (Optional) Time discretization order (defaults to 2).
     :param dimension: :(Optional) class:`Dimension` object for
-                       representing the number of points in this source
-
-    Note, either the dimensions `ntime` and `npoint` or the fully
-    initialised `data` array need to be provided.
+                       representing the number of points in this source.
     """
 
-    def __new__(cls, name, grid, ntime=None, npoint=None, data=None,
-                coordinates=None, **kwargs):
+    def __new__(cls, name, grid, time_range, npoint=None,
+                data=None, coordinates=None, **kwargs):
         p_dim = kwargs.get('dimension', Dimension(name='p_%s' % name))
         time_order = kwargs.get('time_order', 2)
         npoint = npoint or coordinates.shape[0]
-        if data is None:
-            if ntime is None:
-                error('Either data or ntime are required to'
-                      'initialise source/receiver objects')
-        else:
-            ntime = ntime or data.shape[0]
+
+        cls._time_range = TimeAxis(time_range.start, time_range.step, time_range.num)
```

Comment:
are you just rebuilding the ``time_range``? if so, perhaps it's worth adding a copy/rebuild method to ``TimeAxis`` itself. Also, u shouldn't probably attach this to ``cls``, but rather to ``obj`` directly

Reply:
fixed in e2024c50